### PR TITLE
build: generate pkgconfig file

### LINKED ISSUE #472: build: write a pkg-config file
This is a request to enable writing out a `pkg-config` file when `blis` is built. This would make it easy for `CMake` projects that use `blis` to [find](https://cmake.org/cmake/help/latest/module/FindPkgConfig.html) and incorporate the `blis` configuration (`Cflags`, `libs`) appropriately. Thanks!

### PR DESCRIPTION
**Pull request by ajaypanyala**
Fixes #472

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -166,6 +166,7 @@ MK_INCL_DIR_INST          := $(INSTALL_INCDIR)/blis
 # Set the path to the subdirectory of the share installation directory.
 MK_SHARE_DIR_INST         := $(INSTALL_SHAREDIR)/blis
 
+PC_SHARE_DIR_INST         := $(INSTALL_SHAREDIR)/pkgconfig
```
Comment (isuruf) [inline Makefile:169]: Isn't this supposed to be `$(INSTALL_LIBDIR)/pkgconfig`?

Comment (ajaypanyala) [reply Makefile:169]: It could be, but since the rest of the `blis` config files are installed in `share`, I used the same to be consistent.

Comment (isuruf) [reply Makefile:169]: pkg-config will not find `blis` if it's not installed into `$(INSTALL_LIBDIR)/pkgconfig` right? (Unless the user manually sets `PKG_CONFIG_PATH`)

Comment (ajaypanyala) [reply Makefile:169]: Right. Thanks for bringing that up. I was only testing it from within a `CMake` project and that works either way with `CMake`. How about adding a `libdir=@prefix@/share` to `blis.pc.in`. It should automatically pick up from `datadir/pkgconfig`  as well right ? Alternatively, I could also change it to `$(INSTALL_LIBDIR)/pkgconfig`, but I would prefer @devinamatthews comment on this one before I do that.

**Comment by devinamatthews** [comment]
@ajaypanyala @isuruf I'm in over my head here but [this guide](https://people.freedesktop.org/~dbn/pkg-config-guide.html) says "On a typical Unix system, it will search in the directories /usr/lib/pkgconfig and /usr/share/pkgconfig". On my mac they're all in /usr/lib. On our cluster they're all in /usr/share... Since a pkg-config file is not actually an executable library I would put it in share, but then again .la files go in lib so I guess anything goes.

My main concern is actually using the BLIS CFLAGS and LDFLAGS... these should only be necessary for downstream application to use when linking the static version, and even then not all of the flags are necessary. Perhaps the initial version can assume dynamic linking and omit those?

File: Makefile (hunk 1 above)
Comment (devinamatthews) [reply Makefile:169]: See comment [here](https://github.com/flame/blis/pull/473#issuecomment-758802730).

**Comment by ajaypanyala** [comment]
@devinamatthews I was not sure how to separate `CFLAGS` needed by downstream application from the ones that are private to `blis` build when writing them out to the `pc` file. Since the default `blis` configuration is building both `static` and `dynamic` libs, it might be better not to omit them since we do not know what the downstream application would prefer to use ?

**Comment by isuruf** [comment]
For libraries that are needed only during static linking you can add them to `Libs.private` instead of `Libs` which is used when `pkg-config --static` is used.

**Comment by isuruf** [comment]
> @ajaypanyala @isuruf I'm in over my head here but this guide says "On a typical Unix system, it will search in the directories /usr/lib/pkgconfig and /usr/share/pkgconfig". On my mac they're all in /usr/lib. On our cluster they're all in /usr/share... Since a pkg-config file is not actually an executable library I would put it in share, but then again .la files go in lib so I guess anything goes.

Didn't know that. I'm fine with either option then.

**Comment by ajaypanyala** [comment]
> For libraries that are needed only during static linking you can add them to `Libs.private` instead of `Libs` which is used when `pkg-config --static` is used.

Right. I was not sure how to separate the private ones from the ones needed by downstream apps without intrusive changes to the `blis` build which I do not completely understand. Maybe @devinamatthews can do that.

**Comment by devinamatthews** [comment]
I would say leave them out, and we can put a caveat in the docs that says for now it (pkg-config) only works with dynamic linking. It's still a net gain in functionality.

**Comment by ajaypanyala** [comment]
@devinamatthews I pushed a small change as suggested by @isuruf . With this `pkg-config --libs blis` assumes you want the dynamic version and just gives `-lblis`. A call to `pkg-config --libs blis --static` would return the `Libs.private` as well which could be for example `-lm -lpthread -fopenmp -lrt`.

**Comment by ajaypanyala** [comment]
> My main concern is actually using the BLIS CFLAGS and LDFLAGS... these should only be necessary for downstream application to use when linking the static version, and even then not all of the flags are necessary. 

@devinamatthews Could you point me to a `blis` configuration that would populate `CFLAGS` where some flags are not needed even for static libs. I did not see any such case atleast in the configurations I tested.

**Comment by devinamatthews** [comment]
I can't say for sure, maybe -funsafe-math-optimizations. But even basic optimization flags shouldn't be passed on to downstream applications. If -fopenmp is in Libs.private then I think that is sufficient and CFLAGS can be dropped. Good work on the --static workaround.

**Comment by ajaypanyala** [comment]
Dropped `CFLAGS`. Thanks!

**Comment by ajaypanyala** [comment]
@devinamatthews @isuruf Is this good to go ? Please let me know if something else is needed before merging.

**Comment by devinamatthews** [comment]
Yes, looks good. I'll merge.

**Pull request merged**